src/bio.c

```diff
@@ -278,6 +293,158 @@ void *bioProcessBackgroundJobs(void *arg) {
             if (job_type == BIO_CLOSE_AOF) close(job->fd_args.fd);
         } else if (job_type == BIO_LAZY_FREE) {
             job->free_args.free_fn(job->free_args.free_args);
+        } else if (job_type == BIO_SAVE_TO_DISK) {
+            int usemark;
+            char lastbytes[RDB_EOF_MARK_SIZE], buf[PROTO_IOBUF_LEN], eofmark[RDB_EOF_MARK_SIZE];
+            ssize_t readlen, nread, nwritten;
+            off_t left = 0, repl_transfer_last_fsync_off = 0;
+            int error = 0, main_thread_abort = 0, received_ping = 0;
+            connection *conn = job->save_to_disk_args.conn;
+            int dest_fd = job->save_to_disk_args.dest_fd;
+
+            /* Put the socket in blocking mode to simplify RDB transfer.
+             * We'll restore it when the RDB is received. */
+            connBlock(conn);
+            connRecvTimeout(conn, server.repl_syncio_timeout * 1000);
+            do {
+                received_ping = 0;
+                if (shouldAbortSave()) {
+                    main_thread_abort = 1;
+                    goto done;
+                }
+                nread = connSyncReadLine(conn, buf, 1024, server.repl_syncio_timeout * 1000);
+                if (nread == -1) {
+                    replicaBioSaveServerLog(LL_WARNING, "I/O error reading bulk count from PRIMARY: %s", connGetLastError(conn));
+                    error = 1;
+                    goto done;
+                } else {
+                    /* nread here is returned by connSyncReadLine(), which calls syncReadLine() and
+                     * convert "\r\n" to '\0' so 1 byte is lost. */
+                    server.bio_stat_net_repl_input_bytes += nread + 1;
+                }
+                int ret = inspectBulkPayloadHeaderForErrors(buf);
+                if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_ABORT) {
+                    replicaBioSaveServerLog(LL_WARNING, "PRIMARY aborted replication with an error: %s", buf + 1);
+                    error = 1;
+                    goto done;
+                } else if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_PING) {
+                    atomic_store_explicit(&server.repl_transfer_lastio, atomic_load_explicit(&server.unixtime, memory_order_relaxed), memory_order_relaxed);
+                    memset(buf, 0, PROTO_IOBUF_LEN);
+                    received_ping = 1;
+                } else if (ret == INSPECT_BULK_PAYLOAD_PRIMARY_BAD_PROTO) {
+                    replicaBioSaveServerLog(LL_WARNING,
+                                            "Bad protocol from PRIMARY, the first byte is not '$' (we received '%s'), are you sure the host "
```

Comment:
This log (and the one above) is repeated in multiple places. Put it in a shared constant to avoid duplication and inconsistency.